sdk/core/azure-core/inc/azure/core/context.hpp

```diff
@@ -233,7 +233,17 @@ namespace Azure { namespace Core {
      * @brief Checks if the context is cancelled.
      * @return `true` if this context is cancelled; otherwise, `false`.
      */
-    bool IsCancelled() const noexcept { return GetDeadline() < std::chrono::system_clock::now(); }
+    bool IsCancelled() const noexcept
+    {
+      try
+      {
+        return GetDeadline() < std::chrono::system_clock::now();
```

Comment:
I don't think it's a good idea to do a localized try-catch like this, to fix the issue.

Either IsCancelled is designed to be no except, in which case we should validate and mark GetDeadline() as no except, or we need to re-think IsCancelled as a no-except method.

Reply:
It felt icky to me as well. But I could easily imagine scenarios where IsCancelled could be called from a destructor, which would require that it be noexcept.

Reply:
Oh, I see now (catching up from above conversation), marking GetDeadline as no-except requires the ctor to be marked as well, which we can't enforce :(

As @vhvb1989 mentioned, we should evaluate where to use noexcept and whether it makes sense to remove it here.

Conceptually, IsCancelled was intended to be a noexcept API. I wish we could guarantee that "truly" based on the implementation. If we can't, for such a method, it might be difficult to guarantee for other more complex methods.

Reply:
> Conceptually, IsCancelled was intended to be a noexcept API

`IsCancelled()` can be `noexcept` if all it does is to check some internal state, like

```cpp
// returns an internal state of the context
return m_cancelled;
```

However, we are using `IsCancelled()` as a trigger for a validation like `shouldCancelNow()` or `isTimeOut()`.  A trigger like this should update the internal state and set the Context as `Cancelled`.

So, we can fix this by adding a new API method that triggers the validation, and update `IsCancelled()` implementation to only return the current encapsulated state.

Otherwise, `IsCancelled()` should not be `noexcept` as it is today, because it depends on a chain of events that could change in the future.